src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+                        }
+
+                        var task = InvokeNextResultFilterAsync();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ResultSyncEnd;
+                            return task;
+                        }
+
+                        goto case State.ResultSyncEnd;
+                    }
+
+                case State.ResultSyncEnd:
+                    {
+                        Debug.Assert(_resultExecutedContext != null);
+
+                        var filter = (IResultFilter)state;
+                        var resultExecutedContext = _resultExecutedContext;
 
-                        item.Filter.OnResourceExecuted(_resourceExecutedContext);
+                        _diagnosticSource.BeforeOnResultExecuted(resultExecutedContext, filter);
 
-                        _diagnosticSource.AfterOnResourceExecuted(_resourceExecutedContext, item.Filter);
+                        filter.OnResultExecuted(resultExecutedContext);
+
+                        _diagnosticSource.AfterOnResultExecuted(resultExecutedContext, filter);
+
+                        goto case State.ResultEnd;
                     }
-                }
-                else
-                {
-                    // >> ExceptionFilters >> Model Binding >> ActionFilters >> Action
-                    await InvokeAllExceptionFiltersAsync();
 
-                    // If Exception Filters provide a result, it's a short-circuit due to an exception.
-                    // We don't execute Result Filters around the result.
-                    Debug.Assert(_exceptionContext != null);
-                    if (_exceptionContext.Result != null)
+                case State.ResultInside:
                     {
-                        // This means that exception filters returned a result to 'handle' an error.
-                        // We're not interested in seeing the exception details since it was handled.
-                        await InvokeResultAsync(_exceptionContext.Result);
+                        // The empty result is always flowed back as the 'executed' result
+                        var result = _resultExecutingContext?.Result ?? _result ?? new EmptyResult();
+                        _result = result;
 
-                        _resourceExecutedContext = new ResourceExecutedContext(_resourceExecutingContext, _filters)
+                        var task = InvokeResultAsync(_result);
+                        if (task.Status != TaskStatus.RanToCompletion)
                         {
-                            Result = _exceptionContext.Result,
-                        };
+                            next = State.ResultEnd;
+                            return task;
+                        }
+
+                        goto case State.ResultEnd;
                     }
-                    else if (_exceptionContext.Exception != null)
+
+                case State.ResultEnd:
```

Comment:
It would be good to add Debug.Assert(scope != <<scope>>) for some of these states, like in this case maybe Debug.Assert(scope != Scope.Exception). Also on the parts where the stack gets "Winded" and "Unwinded" its worth pointing out what happens with the state transitions, basically what the scenario is that we are covering on each of the branches. i.e, for scope == Scope.Resource indicate that the action being invoked contained resource filters and that we are unwinding those now.